From c38d135826147ff5a299fcd1ae2398201be71638 Mon Sep 17 00:00:00 2001
From: Martin Pitt <martin.pitt@ubuntu.com>
Date: Wed, 24 Mar 2010 09:16:22 +0100
Subject: [PATCH] keyboard: Fix layout gconf key initialization for empty variants

There is currently an inconsistency in how g-keyboard-properties and
gdm/g-s-d's $GDM_KEYBOARD_LAYOUT handle keyboard layouts without variants.
gdm/gsd create them just like "us", while g-k-p creates them as "us\t" after
reordering.

This causes duplication of layouts when restarting g-s-d, since that will then
add another "us" (since $GDM_KEYBOARD_LAYOUT == "us").

Bug: https://bugzilla.gnome.org/show_bug.cgi?id=613775
Bug-Ubuntu: https://launchpad.net/bugs/460328
---
 .../keyboard/gnome-keyboard-properties-xkblt.c     |   14 ++++++++++----
 1 files changed, 10 insertions(+), 4 deletions(-)

Index: gnome-control-center-2.31.90/capplets/keyboard/gnome-keyboard-properties-xkblt.c
===================================================================
--- gnome-control-center-2.31.90.orig/capplets/keyboard/gnome-keyboard-properties-xkblt.c	2010-08-17 02:46:46.000000000 +1000
+++ gnome-control-center-2.31.90/capplets/keyboard/gnome-keyboard-properties-xkblt.c	2010-09-01 17:39:26.502617000 +1000
@@ -95,12 +95,18 @@
 					GCONF_VALUE_STRING, NULL);
 	if (retval == NULL) {
 		GSList *cur_layout;
+		gchar *layout;
+		int len;
 
 		for (cur_layout = initial_config.layouts_variants;
-		     cur_layout != NULL; cur_layout = cur_layout->next)
-			retval =
-			    g_slist_prepend (retval,
-					     g_strdup (cur_layout->data));
+		     cur_layout != NULL; cur_layout = cur_layout->next) {
+		        layout = g_strdup (cur_layout->data);
+			/* chop off empty variants to avoid duplicates */
+			len = strlen (layout);
+			if (layout[len - 1] == '\t')
+				layout[len - 1] = '\0';
+			retval = g_slist_prepend (retval, layout);
+		    }
 
 		retval = g_slist_reverse (retval);
 	}
